### PR TITLE
[IM] Rewrite `_object_type_validator()` and deprecate `_dict_type_validator()`

### LINKED ISSUE #1401: [IM] Reviewing recent changes to IM and making plans for housekeeping tasks
This issue mostly focuses on figuring out what housekeeping tasks that IM would need to improve further. Due to some significant additions to the IM (i.e., mofidiable data at runtime, array validation, and unifying json and csv formats), it is good to take a step back and review how those new changes fit together and what can be streamlined and refactored.

- [x] Remove the `_dict_type_validator()` method and use `_validate_input_by_type()` directly in `_populate_pool()` and `_add_variable_to_pool()`, and fix broken tests.
- [x] Check test coverage and improve it.
- [ ] Review the interaction between the two features - modifiability at runtime and array validation.
- [ ] Review the `called_during_initialization` parameter.
- [ ] Review the usage of the ElementsCounter and ElementState classes to see if it should be integrated into the InputManager class.
- [ ] `_add_variable_to_pool()` is too complex according to flake8.
- [ ] ...

### LINKED ISSUE #1409: [IM][Refactor] Simplify the use of the object/dictionary validation in `_populate_pool()` and `_add_variable_to_pool()`
## Urgency Level

- [ ] Urgent: a critical feature or fix is needed
- [ ] High: the issue significantly impacts the user experience
- [x] Medium: the issue affects functionality but has a workaround
- [ ] Low: the issue doesn't have a significant immediate impact
- [ ] Wishful: the issue should not be prioritized over any others
- [ ] Unsure: the urgency level has not been or cannot be assessed at this time 

## Status

The two methods `_populate_pool()` and `_add_variable_to_pool()` are currently using `_dict_type_validator()`, which is fairly similar to another method `_object_type_validator()`. 

## Intention

So, it would be good to remove the `_dict_type_validator()` method and funnel everything through the `_validate_input_by_type()` instead, which will then make subcalls to different type validators. That method will call the `_object_type_validator()` when appropriate.

## Request

- [x] Replace `_dict_type_validator()` in `_populate_pool()` with `_validate_input_by_type()`.
- [x] Replace `_dict_type_validator()` in `_add_variable_to_pool()` with `_validate_input_by_type()`.
- [x] Fix/Replace broken unit tests.

### PR DESCRIPTION
**Pull request by allisterakun**
Rewrite `_object_type_validator()` to iterate through the metadata and deprecate `_dict_type_validator()`.

### Context

Issue(s) closed by this pull request: closes #1401 #1409 

- [x] The [changelog](https://docs.google.com/spreadsheets/d/1U-3igxdstHTFb6k5dVcE5-x9IL0r6q18eUiwwytKZyA/edit#gid=0) has been updated.

### What

1. Rewrite `_object_type_validator()` to iterate through metadata instead of input data for data validation.
2. Deprecate `_dict_type_validator()`, and rerout all occurrences to directly call `_validate_input_by_type()` instead.
3. Update all affected unit tests.
4. Bring test coverage for IM to 100%.

### Why

1. When validating data, we should loop through all the properties present in the metadata and extract the corresponding variable in the input data, not the other way around.
2. As mentioned in #1401 and #1409, `_dict_type_validator()` is an unnecessary wrapper function to call `_validate_input_by_type()`.

### How

1. Update the for loop in `_object_type_validator()` to iterate through the metadata properties
2. Reroute all occurrences of `_dict_type_validator()` to call '_validate_input_by_type()' instead.

### Test plan

All affected unit tests are updated, and added 3 more tests to bring the test coverage for IM to 100%.

### CODE REVIEW

**Review by tomhuhh — APPROVED** [review]
Thanks Aliister! I tested with a few scenarios including invalid `male calf ratio`, `total solids removal efficiency`, `nitrogen removal efficiency`, and `potassium removal efficiency`. The warning system seems to work well for those scenarios. 

Great work, Allister!

File: README.md [hunk 1]
```diff
@@ -1,7 +1,7 @@
 [![Testing](https://github.com/RuminantFarmSystems/MASM/actions/workflows/testing_pytest.yml/badge.svg)](https://github.com/RuminantFarmSystems/MASM/actions/workflows/testing_pytest.yml)
 [![Linting](https://github.com/RuminantFarmSystems/MASM/actions/workflows/lint_flake8.yml/badge.svg)](https://github.com/RuminantFarmSystems/MASM/actions/workflows/lint_flake8.yml)
 [![Coverage](https://img.shields.io/badge/coverage-95%25-brightgreen)](https://github.com/RuminantFarmSystems/MASM/actions/workflows/coverage_pytest-cov.yml)
-[![Mypy](https://img.shields.io/badge/mypy-4427%20errors-red)](https://github.com/RuminantFarmSystems/MASM/actions/workflows/mypy_type_checker.yml)
+[![Mypy](https://img.shields.io/badge/mypy-4446%20errors-red)](https://github.com/RuminantFarmSystems/MASM/actions/workflows/mypy_type_checker.yml)
```
Comment (Isgotkowitz) [inline README.md]: Would it be possible to address some of these Mypy errors so the total number doesn't increase? I'm not sure how involved they would be to fix, but in my experience there are usually some pretty trivial fixes for them. `check_changes.sh` helps to filter them out and identify the lowest hanging fruit.

**Review by Isgotkowitz — APPROVED** [review]
LGTM, nice job Allister. I tried with all the metadata files, and 3 of them don't work in their current state (barnyard, SWAT, and testing). But it only takes changing their feed file to "default_feed.json" to fix them, so I think it isn't a big deal.

File: README.md (hunk 1 above)
Comment (allisterakun) [reply README.md]: I will take a look! Thank you!

**Pull request merged**